src/validation-engine.ts

```diff
@@ -0,0 +1,238 @@
+import {
+  ValidationResult,
+  FormValidationResult,
+  FieldsValidationSchema,
+  RecordValidationFunction,
+  createDefaultValidationResult,
+  RecordValidationFunctionSyncAsync,
+  isSyncValidationResult,
+  FieldValidationFunctionSyncAsync,
+  FieldValidationFunction,
+  FieldValidation,
+} from './model';
+
+import { isUndefinedOrNull } from './helper';
+
+import {
+  fireAllFieldsValidations,
+  fireSingleFieldValidations,
+  fireRecordValidations,
+} from './validation-dispatcher';
+import { buildFormValidationResult } from './form-validation-summary-builder';
+
+export class ValidationEngine {
+  private validationsPerField: FieldsValidationSchema = {};
+  private validationsGlobalForm: RecordValidationFunction[] = [];
+  private asyncValidationInProgressCount = 0;
+
+  // TODO, in new api we are adding an optional parameter
+  // to customize the error message (string array)
+  //
+  // TODO: Chain pattern applied here, check why it is useful
+  addFieldValidation(
+    key: string,
+    validation: FieldValidationFunctionSyncAsync,
+    customParams: any = {},
+    errorMessage?: string
+  ): ValidationEngine {
+    const asyncValidationFn = this.convertFieldValidationToAsyncIfNeeded(
+      validation
+    );
+
+    if (!this.isFieldKeyMappingDefined(key)) {
+      this.validationsPerField[key] = [];
+    }
+
+    const fieldValidation: FieldValidation = {
+      validator: asyncValidationFn,
+      customArgs: customParams,
+    };
+
+    if (errorMessage) {
+      fieldValidation.errorMessage = errorMessage;
+    }
+
+    this.validationsPerField[key].push({
+      validator: asyncValidationFn,
+      customArgs: customParams,
+      errorMessage: errorMessage,
+    });
+    return this;
+  }
+
+  // TODO: Should it be moved to model?
```

Comment:
I think, it's like a "helper" or "mapper" function, not model, isn't it?

It's like "convert to async result" or "map to async result"

Reply:
should we create a file called mappers?